misc/dbt-materialize/dbt/include/materialize/macros/is_cluster_ready.sql

```diff
@@ -0,0 +1,140 @@
+-- Copyright Materialize, Inc. and contributors. All rights reserved.
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License in the LICENSE file at the
+-- root of this repository, or online at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+{% macro is_cluster_ready(cluster=target.cluster|default(none)) %}
+
+-- Check if the cluster argument was provided or exists in the target profile
+{% if cluster is none %}
+    {{ exceptions.CompilationError("No cluster specified and no default cluster found in target profile. " ~ current_target_name) }}
+{% endif %}
+
+{{ log("Checking pending objects for cluster " ~ cluster, info=True) }}
+
+{% set cluster_exists %}
+WITH total_replicas AS (
+    SELECT cluster_id, count(*) AS replicas
+    FROM mz_cluster_replicas
+    GROUP BY cluster_id
+)
+
+SELECT COALESCE(replicas, 0) AS replicas
+FROM mz_clusters
+LEFT JOIN total_replicas ON mz_clusters.id = cluster_id
+WHERE mz_clusters.name = lower('{{ cluster }}')
+{%- endset -%}
+
+{%- set result = run_query(cluster_exists) %}
+{%- if execute -%}
+    {%- if result and result.rows|length > 0 -%}
+        {%- set replicas_value = result.rows[0][0] %}
+        {%- if replicas_value is none or replicas_value == 0 -%}
+            {{ log("Cluster " ~ cluster ~ " has no running replicas", info=true) }}
+            {{ return(false) }}
+        {%- endif -%}
+    {%- else -%}
+        {{ log("Cluster " ~ cluster ~ " does not exist", info=true) }}
+        {{ return(false) }}
+    {%- endif -%}
+{%- endif -%}
+
+{%- set check_pending_objects_sql %}
```

Comment:
I think this could be simplified. I'm assuming the intent is to find out the point at which:
* all objects on this cluster are hydrated
* objects that have inputs ("materializations") have a small lag.

It should be possible to use `mz_hydration_statuses` and `mz_materialization_lag` for this. The former contains an entry for all objects that are backed by dataflows (indexes, MVs, sources, sinks), the latter contains an entry for all objects that have inputs. So what you could do is:

* For all indexes, MVs, sources, and sinks...
    * Check if they are `hydrated` in `mz_hydration_statuses`.
    * If they have an entry in `mz_materialization_lag`, check if `local_lag` is <= 1s.

One thing to note is that both `mz_hydration_statuses` and `mz_materialization_lag` are per replica, not per cluster. I think to handles this correctly we could perform the check by replica and then say "a cluster is ready if at least one replica is ready".

Reply:
You're right. That's much simpler 😄